### PR TITLE
whitelist + topwallet through the admin proxy contract

### PR DESCRIPTION
**Pull request by StanislavShevchenko**
# Description

whitelist + topwallet through the admin proxy contract

set to env `NUMBER_OF_ADMIN_WALLET_ACCOUNTS=1`

About #100 

# How Has This Been Tested?

registration a new user 

# Checklist:
- [x] PR title matches follow: (Feature|Bug|Chore) Task Name
- [x] My code follows the style guidelines of this project
- [x] I have followed all the instructions described in the initial task (check Definitions of Done)
- [x] I have performed a self-review of my own code
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have added reference to a related issue in the repository
- [x] I have added a detailed description of the changes proposed in the pull request. I am as descriptive as possible, assisting reviewers as much as possible.
- [ ] I have added screenshots related to my pull request ( for frontend tasks)
- [ ] I have pasted a gif showing the feature.
- [ ] @mentions of the person or team responsible for reviewing proposed changes

### CODE REVIEW

File: src/server/blockchain/AdminWallet.js [hunk 1]
```diff
@@ -358,39 +362,16 @@ export class Wallet {
...
-          from: this.address,
-          to: address,
-          value: toTop,
-          gas: defaultGas,
-          gasPrice: defaultGasPrice
-        })
-        log.debug('Topwallet result:', { res })
-        return res
+  async topWallet(address: string): PromiEvent<TransactionReceipt> {
+    const tx: TransactionReceipt = await this.sendTransaction(this.proxyContract.methods.topWallet(address)).catch(
```
Comment (sirpy) [inline src/server/blockchain/AdminWallet.js]: why were all the checks removed?

File: src/server/blockchain/AdminWallet.js [hunk 2]
```diff
@@ -289,21 +294,20 @@ export class Wallet {
   /**
    * whitelist an user in the `Identity` contract
    * @param {string} address
-   * @param {string} did
    * @returns {Promise<TransactionReceipt>}
    */
-  async whitelistUser(address: string, did: string): Promise<TransactionReceipt | boolean> {
+  async whitelistUser(address: string): Promise<TransactionReceipt | boolean> {
```
Comment (sirpy) [inline src/server/blockchain/AdminWallet.js]: that's a mistake did should not have been removed... we need to fix the adminwallet contracts

Comment (StanislavShevchenko) [reply src/server/blockchain/AdminWallet.js]: done in https://github.com/GoodDollar/GoodServer/pull/139/commits/113e7728f6d745c5e872c0ddfb275e9e9390919a

File: src/server/blockchain/AdminWallet.js (hunk 1 above)
Comment (StanislavShevchenko) [reply src/server/blockchain/AdminWallet.js]: done in https://github.com/GoodDollar/GoodServer/pull/139/commits/113e7728f6d745c5e872c0ddfb275e9e9390919a

Comment (StanislavShevchenko) [reply src/server/blockchain/AdminWallet.js]: > why were all the checks removed?

I returned all checks

File: src/server/blockchain/AdminWallet.js [hunk 3]
```diff
@@ -375,12 +384,9 @@ export class Wallet {
...
       log.debug('TopWallet:', { userBalance, toTop })
       if (toTop > 0 && (force || toTop / maxTopWei >= 0.75)) {
-        let res = await this.sendNative({
-          from: this.address,
-          to: address,
-          value: toTop,
-          gas: defaultGas,
-          gasPrice: defaultGasPrice
+        let res = await this.sendTransaction(this.proxyContract.methods.topWallet(address)).catch(e => {
+          log.error('Error topWallet', { e }, e.message, { address })
```
Comment (sirpy) [inline src/server/blockchain/AdminWallet.js]: redundant catch

File: src/server/blockchain/__tests__/adminWallet.test.js [hunk 4]
```diff
@@ -48,7 +48,7 @@ describe('adminwallet', () => {
     const promises = []
     for (let i = 0; i < 5; i++) {
       const unverifiedAddress = generateWalletAddress()
-      // console.log('unverifiedAddress', unverifiedAddress)
+      const tx = await AdminWallet.whitelistUser(unverifiedAddress, 'did:gd')
```
Comment (sirpy) [inline src/server/blockchain/__tests__/adminWallet.test.js]: whats this for?

File: src/server/blockchain/__tests__/adminWallet.test.js [hunk 5]
```diff
@@ -57,12 +57,14 @@ describe('adminwallet', () => {
 
   test('adminWallet bad transaction in queue', async () => {
     const unverifiedAddress = generateWalletAddress()
+    const unverifiedAddress2 = generateWalletAddress()
```
Comment (sirpy) [inline src/server/blockchain/__tests__/adminWallet.test.js]: what are these extra checks on address2 for?

Comment (StanislavShevchenko) [reply src/server/blockchain/__tests__/adminWallet.test.js]: The second address is entered, for testing.
topWallet method from smartcontract AdminWallet does not allow to add one address twice

File: src/server/blockchain/__tests__/adminWallet.test.js (hunk 4 above)
Comment (StanislavShevchenko) [reply src/server/blockchain/__tests__/adminWallet.test.js]: I cannot make topWallet without adding to whitelistUser

File: src/server/blockchain/AdminWallet.js (hunk 3 above)
Comment (StanislavShevchenko) [reply src/server/blockchain/AdminWallet.js]: done in https://github.com/GoodDollar/GoodServer/pull/139/commits/1ad3e83aab6fc53447f5709900332c4d7a565d7c

File: src/server/blockchain/__tests__/adminWallet.test.js (hunk 4 above)
Comment (sirpy) [reply src/server/blockchain/__tests__/adminWallet.test.js]: well then obviously its a broken behaviour, the tests are here for THIS PURPOSE EXACTLY

Comment (sirpy) [reply src/server/blockchain/__tests__/adminWallet.test.js]: this should work without whitelist
@AlexeyKosinski

File: src/server/blockchain/__tests__/adminWallet.test.js (hunk 5 above)
Comment (sirpy) [reply src/server/blockchain/__tests__/adminWallet.test.js]: @AlexeyKosinski 
dont understand this

Comment (AlexeyKosinski) [reply src/server/blockchain/__tests__/adminWallet.test.js]: we have check inside smartcontract. if address already whitelisted tx will be reverted

File: src/server/blockchain/__tests__/adminWallet.test.js (hunk 4 above)
Comment (AlexeyKosinski) [reply src/server/blockchain/__tests__/adminWallet.test.js]: so we need to change smartcontract
because topWallet dodn't work without whitelist

Comment (AlexeyKosinski) [reply src/server/blockchain/__tests__/adminWallet.test.js]: also need to add `did` param to AdminWallet.whitelistUser

**Comment by AlexeyKosinski** [comment]
we have problem with run a lot of txes (methods in contract) in one blockchain block.

example:
when I run

TEST1
```
AdminWallet.topWallet(addr1)
AdminWallet.topWallet(addr2)
```

second tx will be reverted
Screenshot 2020-01-23 at 15 31 59

but when I set delay between

TEST2
```
AdminWallet.topWallet(addr1)
await delay(800)
AdminWallet.topWallet(addr2)
```
all txes success.
Screenshot 2020-01-23 at 15 30 53

I’ve investigated logs I saw only one difference
in TEST1 txes mined in the same block
in TEST2 txes mined in different blocks

the same if we use another contract function. for ex
AdminWallet.whitelistUser
`so second call of the same function in one block will be reverted`

**Pull request merged**